pkgs/development/libraries/libxcrypt/default.nix

```diff
@@ -28,8 +28,20 @@ stdenv.mkDerivation (finalAttrs: {
     "--disable-werror"
   ];
 
-  # fixes: can't build x86_64-w64-mingw32 shared library unless -no-undefined is specified
-  makeFlags = lib.optionals stdenv.hostPlatform.isWindows [ "LDFLAGS=-no-undefined"] ;
+  makeFlags = let
+    lld17Plus = stdenv.cc.bintools.isLLVM
+      && lib.versionAtLeast stdenv.cc.bintools.version "17";
+
+    linkFlags = []
+      # fixes: can't build x86_64-w64-mingw32 shared library unless -no-undefined is specified
+      ++ lib.optionals stdenv.hostPlatform.isWindows [ "-no-undefined" ]
+
+      # lld 17 sets `--no-undefined-version` by default and `libxcrypt`'s
+      # version script unconditionally lists legacy compatibility symbols, even
+      # when not exported: https://github.com/besser82/libxcrypt/issues/181
+      ++ lib.optionals lld17Plus ["-Wl,--undefined-version"]
+    ;
+  in lib.optionals (linkFlags != []) [ "LDFLAGS='${toString linkFlags}'" ];
```

Comment:
i don't think that `LDFLAGS` can contain spaces when passed via `makeFlags` without `__structuredAttrs`, tho also not sure if it is possible to have `isWindows` true with `cc.bintools.isLLVM`

Reply:
> i don't think that `LDFLAGS` can contain spaces when passed via `makeFlags` without `__structuredAttrs`,

ah, good catch! I misunderstood how quoting works for those array variables; that does indeed seem to be the case:

```console
nix-repl> :b np.stdenv.mkDerivation { makeFlags = ["LDFLAGS='foo bar baz'"]; name = ""; dontUnpack = true; configurePhase = ''make() { for a in "$@"; do echo $a; done; }''; installPhase = "exit 1"; }
error: builder for '/nix/store/l23n1s685p6qn6mzd0m7bghjp9zy3rha-unknown.drv' failed with exit code 1;
       last 10 log lines:
       > patching sources
       > updateAutotoolsGnuConfigScriptsPhase
       > configuring
       > building
       > build flags: SHELL=/nix/store/z3f9pjbakmf26p4081y6h046fsn4yvq3-bash-5.2-p15/bin/bash LDFLAGS=\'foo bar baz\'
       > SHELL=/nix/store/z3f9pjbakmf26p4081y6h046fsn4yvq3-bash-5.2-p15/bin/bash
       > LDFLAGS='foo
       > bar
       > baz'
       > installing
       For full logs, run 'nix log /nix/store/l23n1s685p6qn6mzd0m7bghjp9zy3rha-unknown.drv'.
```

> tho also not sure if it is possible to have `isWindows` true with `cc.bintools.isLLVM`

I haven't kept up with it but IIRC there's been ongoing work on a LLVM based stdenv targeting mingwW64.

Reply:
Pushed a fix using `LDFLAGS+=`, PTAL